tests/unit/s2n_random_test.c

```diff
@@ -184,169 +111,338 @@ int main(int argc, char **argv)
+    EXPECT_OK(s2n_fork_test(s2n_get_public_random_data, GET_PRIVATE_RANDOM_DATA));
+    EXPECT_OK(s2n_fork_test(s2n_get_private_random_data, GET_PUBLIC_RANDOM_DATA));
+
+    /* Basic tests generating randomness */
+    EXPECT_OK(s2n_basic_generate_tests());
+
+    /* Verify that there are no trivially observable patterns in the output */
+    EXPECT_OK(s2n_basic_pattern_tests(s2n_get_public_random_data));
+    EXPECT_OK(s2n_basic_pattern_tests(s2n_get_private_random_data));
+
+    /* Just a sanity check and avoids cppcheck "unassignedVariable" errors.
+     * In future PRs this part will be expanded.
+     */
+    blob1.size = RANDOM_GENERATE_DATA_SIZE;
+    EXPECT_OK(s2n_get_public_random_data(&blob1));
+    blob2.size = RANDOM_GENERATE_DATA_SIZE;
+    EXPECT_OK(s2n_get_private_random_data(&blob2));
+    EXPECT_BYTEARRAY_NOT_EQUAL(data1, data2, RANDOM_GENERATE_DATA_SIZE);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_random_test_case_default_cb(struct random_test_case *test_case)
+{
+    EXPECT_SUCCESS(s2n_init());
+
+    /* Verify that randomness callbacks can't be set to NULL */
+    EXPECT_FAILURE(s2n_rand_set_callbacks(NULL, s2n_cleanup_cb, s2n_entropy_cb, s2n_entropy_cb));
+    EXPECT_FAILURE(s2n_rand_set_callbacks(s2n_init_cb, NULL, s2n_entropy_cb, s2n_entropy_cb));
+    EXPECT_FAILURE(s2n_rand_set_callbacks(s2n_init_cb, s2n_cleanup_cb, NULL, s2n_entropy_cb));
+    EXPECT_FAILURE(s2n_rand_set_callbacks(s2n_init_cb, s2n_cleanup_cb, s2n_entropy_cb, NULL));
+
+    EXPECT_EQUAL(s2n_common_tests(test_case), S2N_SUCCESS);
+
+    EXPECT_SUCCESS(s2n_cleanup());
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_random_test_case_without_pr_cb(struct random_test_case *test_case)
+{
+    EXPECT_SUCCESS(s2n_init());
+
+    POSIX_GUARD_RESULT(s2n_ignore_prediction_resistance_for_testing(true));
+    EXPECT_EQUAL(s2n_common_tests(test_case), S2N_SUCCESS);
+    POSIX_GUARD_RESULT(s2n_ignore_prediction_resistance_for_testing(false));
+
+    EXPECT_SUCCESS(s2n_cleanup());
+
+    return S2N_SUCCESS;
+}
+
+#define NUMBER_OF_RANDOM_TEST_CASES 2
+struct random_test_case random_test_cases[NUMBER_OF_RANDOM_TEST_CASES] = {
+    {"Random API.", s2n_random_test_case_default_cb, CLONE_TEST_DETERMINE_AT_RUNTIME},
+    {"Random API without prediction resistance.", s2n_random_test_case_without_pr_cb, CLONE_TEST_DETERMINE_AT_RUNTIME},};
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST_NO_INIT();
+
+    EXPECT_TRUE(s2n_array_len(random_test_cases) == NUMBER_OF_RANDOM_TEST_CASES);
+
+    /* Create NUMBER_OF_RANDOM_TEST_CASES number of child processes that run
+     * each test case.
+     *
+     * Fork detection is lazily initialised on first invocation of
+     * s2n_get_fork_generation_number(). Hence, it is important that childs are
+     * created before calling into the fork detection code.
+     */
+    pid_t proc_ids[NUMBER_OF_RANDOM_TEST_CASES] = {0};
+
+    for (size_t i = 0; i < NUMBER_OF_RANDOM_TEST_CASES; i++) {
+
+        proc_ids[i] = fork();
```

Comment:
Why this initial fork? Doesn't s2n_fork_test test the fork + pthreads case?

Reply:
Clarified offline: the initial fork isn't necessary currently, but it will be once Torben starts enabling and disabling fork protection features.